File: workorder/server/modules/equipment.py
"""设备管理模块：设备台账 + 点检/保养/维修记录，跟踪设备状态。

约定式插件：``server.api`` 自动应用 :data:`SCHEMA`、挂载 :func:`routes`、调用 :func:`seed`。
业务校验失败抛 :class:`ValidationError`（→400），资源不存在抛 :class:`NotFound`（→404）。
"""

from server.store import ValidationError, NotFound, _now, _num, _require

# 设备状态：运行中 / 闲置 / 保养中 / 故障
EQUIPMENT_STATUS = {
    "running": "运行中",
    "idle": "闲置",
    "maintenance": "保养中",
    "fault": "故障",
}

# 维护记录类型：点检 / 保养 / 维修
LOG_TYPE = {
    "inspect": "点检",
    "maintain": "保养",
    "repair": "维修",
}

SCHEMA = """
CREATE TABLE IF NOT EXISTS equipment (
    id          INTEGER PRIMARY KEY AUTOINCREMENT,
    code        TEXT NOT NULL UNIQUE,
    name        TEXT NOT NULL,
    model       TEXT DEFAULT '',
    location    TEXT DEFAULT '',
    owner       TEXT DEFAULT '',
    status      TEXT NOT NULL DEFAULT 'running',
    created_at  TEXT NOT NULL
);

CREATE TABLE IF NOT EXISTS equipment_logs (
    id            INTEGER PRIMARY KEY AUTOINCREMENT,
    equipment_id  INTEGER NOT NULL REFERENCES equipment(id),
    type          TEXT NOT NULL,
    content       TEXT NOT NULL,
    operator      TEXT DEFAULT '',
    cost          REAL NOT NULL DEFAULT 0,
    created_at    TEXT NOT NULL
);

CREATE INDEX IF NOT EXISTS idx_eqlog_eq ON equipment_logs(equipment_id);
"""


def _equip_label(d: dict) -> dict:
    d["status_label"] = EQUIPMENT_STATUS.get(d["status"], d["status"])
    return d


def _log_label(d: dict) -> dict:
    d["type_label"] = LOG_TYPE.get(d["type"], d["type"])
    return d


def list_equipment(store) -> list[dict]:
    with store._conn() as conn:
        rows = conn.execute("SELECT * FROM equipment ORDER BY id DESC").fetchall()
    return [_equip_label(dict(r)) for r in rows]


def create_equipment(store, data: dict) -> dict:
    code = _require(data, "code", "设备编码")
    name = _require(data, "name", "设备名称")
    status = data.get("status", "running") or "running"
    if status not in EQUIPMENT_STATUS:
        raise ValidationError(f"未知设备状态：{status}")
    with store._lock, store._conn() as conn:
        if conn.execute("SELECT 1 FROM equipment WHERE code = ?", (code,)).fetchone():
            raise ValidationError(f"设备编码 {code} 已存在")
        cur = conn.execute(
            """INSERT INTO equipment (code, name, model, location, owner, status, created_at)
               VALUES (?,?,?,?,?,?,?)""",
            (code, name, str(data.get("model", "")).strip(),
             str(data.get("location", "")).strip(),
             str(data.get("owner", "")).strip(), status, _now()),
        )
        row = conn.execute("SELECT * FROM equipment WHERE id = ?", (cur.lastrowid,)).fetchone()
    return _equip_label(dict(row))


def set_status(store, eid: int, status: str) -> dict:
    if status not in EQUIPMENT_STATUS:
        raise ValidationError(f"未知设备状态：{status}")
    with store._lock, store._conn() as conn:
        if not conn.execute("SELECT 1 FROM equipment WHERE id = ?", (eid,)).fetchone():
            raise NotFound(f"设备 {eid} 不存在")
        conn.execute("UPDATE equipment SET status = ? WHERE id = ?", (status, eid))
        row = conn.execute("SELECT * FROM equipment WHERE id = ?", (eid,)).fetchone()
    return _equip_label(dict(row))


def list_logs(store, equipment_id=None) -> list[dict]:
    sql = ("SELECT l.*, e.name AS equipment_name FROM equipment_logs l "
           "JOIN equipment e ON e.id = l.equipment_id WHERE 1=1")
    params: list = []
    if equipment_id:
        sql += " AND l.equipment_id = ?"
        params.append(equipment_id)
    sql += " ORDER BY l.id DESC"
    with store._conn() as conn:
        rows = conn.execute(sql, params).fetchall()
    return [_log_label(dict(r)) for r in rows]


def add_log(store, eid: int, data: dict) -> dict:
    log_type = _require(data, "type", "记录类型")
    if log_type not in LOG_TYPE:
        raise ValidationError(f"未知记录类型：{log_type}")
    content = _require(data, "content", "记录内容")
    cost = _num(data.get("cost", 0), "费用")
    with store._lock, store._conn() as conn:
        if not conn.execute("SELECT 1 FROM equipment WHERE id = ?", (eid,)).fetchone():
            raise NotFound(f"设备 {eid} 不存在")
        now = _now()
        cur = conn.execute(
            """INSERT INTO equipment_logs
               (equipment_id, type, content, operator, cost, created_at)
               VALUES (?,?,?,?,?,?)""",
            (eid, log_type, content, str(data.get("operator", "")).strip(), cost, now),
        )
        # 维修时把设备置为保养中，完成后用户可再手动改回
        if log_type == "repair":
            conn.execute("UPDATE equipment SET status = 'maintenance' WHERE id = ?", (eid,))
        row = conn.execute(
            "SELECT l.*, e.name AS equipment_name FROM equipment_logs l "
            "JOIN equipment e ON e.id = l.equipment_id WHERE l.id = ?",
            (cur.lastrowid,),
        ).fetchone()
    return _log_label(dict(row))


def routes(store):
    return [
        ("GET",  r"/api/equipment",
         lambda req, p: (200, list_equipment(store))),
        ("POST", r"/api/equipment",
         lambda req, p: (201, create_equipment(store, req["body"]))),
        ("POST", r"/api/equipment/(?P<id>\d+)/status",
         lambda req, p: (200, set_status(store, int(p["id"]), req["body"].get("status", "")))),
        ("GET",  r"/api/equipment/(?P<id>\d+)/logs",
         lambda req, p: (200, list_logs(store, equipment_id=int(p["id"])))),
        ("POST", r"/api/equipment/(?P<id>\d+)/logs",
         lambda req, p: (201, add_log(store, int(p["id"]), req["body"]))),
        ("GET",  r"/api/equipment-logs",
         lambda req, p: (200, list_logs(store))),
    ]


def seed(store):
    """幂等：仅当 equipment 为空时灌入演示设备与一条点检记录。"""
    with store._conn() as conn:
        if conn.execute("SELECT 1 FROM equipment LIMIT 1").fetchone():
            return
    e1 = create_equipment(store, {
        "code": "CNC-01", "name": "数控车床", "model": "CK6140",
        "location": "一号车间", "owner": "张工", "status": "running"})
    create_equipment(store, {
        "code": "INJ-02", "name": "注塑机", "model": "HTF160",
        "location": "二号车间", "owner": "王工", "status": "idle"})
    create_equipment(store, {
        "code": "WELD-03", "name": "焊接机器人", "model": "AR-1440",
        "location": "三号车间", "owner": "赵工", "status": "fault"})
    add_log(store, e1["id"], {
        "type": "inspect", "content": "日常点检：油位正常，导轨润滑良好",
        "operator": "李师傅"})

File: workorder/web/mock/equipment.js
"use strict";
/*
 * 设备管理 mock：复刻 server/modules/equipment.py。
 * 设备台账 + 点检/保养/维修记录，跟踪设备状态。
 * 表：equipment / equipment_logs（与 .py 一致）。
 */
(function () {
  if (!window.Mock) return;
  const M = window.Mock;

  // 设备状态：运行中 / 闲置 / 保养中 / 故障
  const EQUIPMENT_STATUS = {
    running: "运行中",
    idle: "闲置",
    maintenance: "保养中",
    fault: "故障",
  };
  // 维护记录类型：点检 / 保养 / 维修
  const LOG_TYPE = {
    inspect: "点检",
    maintain: "保养",
    repair: "维修",
  };

  function equipLabel(d) {
    d.status_label = EQUIPMENT_STATUS[d.status] || d.status;
    return d;
  }
  function logLabel(d) {
    d.type_label = LOG_TYPE[d.type] || d.type;
    return d;
  }

  // 给日志补 equipment_name（JOIN equipment 的 name）
  function logWithEquip(l) {
    const e = M.db.byId("equipment", l.equipment_id);
    return logLabel(Object.assign({}, l, { equipment_name: e ? e.name : "" }));
  }

  function listEquipment() {
    return M.db.table("equipment").slice().sort((a, b) => b.id - a.id).map((r) => equipLabel(Object.assign({}, r)));
  }

  function createEquipment(data) {
    const code = M.require(data, "code", "设备编码");
    const name = M.require(data, "name", "设备名称");
    const status = data.status || "running";
    if (!EQUIPMENT_STATUS[status]) M.bad("未知设备状态：" + status);
    if (M.db.table("equipment").some((e) => e.code === code)) M.bad(`设备编码 ${code} 已存在`);
    const row = M.db.insert("equipment", {
      code,
      name,
      model: String(data.model || "").trim(),
      location: String(data.location || "").trim(),
      owner: String(data.owner || "").trim(),
      status,
      created_at: M.now(),
    });
    M.db.save();
    return equipLabel(Object.assign({}, row));
  }

  function setStatus(eid, status) {
    if (!EQUIPMENT_STATUS[status]) M.bad("未知设备状态：" + status);
    const e = M.db.byId("equipment", eid);
    if (!e) M.notfound(`设备 ${eid} 不存在`);
    e.status = status;
    M.db.save();
    return equipLabel(Object.assign({}, e));
  }

  function listLogs(equipmentId) {
    let rows = M.db.table("equipment_logs").slice();
    if (equipmentId) rows = rows.filter((l) => l.equipment_id === Number(equipmentId));
    rows.sort((a, b) => b.id - a.id);
    return rows.map(logWithEquip);
  }

  function addLog(eid, data) {
    const logType = M.require(data, "type", "记录类型");
    if (!LOG_TYPE[logType]) M.bad("未知记录类型：" + logType);
    const content = M.require(data, "content", "记录内容");
    const cost = M.num(data.cost === undefined || data.cost === null ? 0 : data.cost, "费用");
    const e = M.db.byId("equipment", eid);
    if (!e) M.notfound(`设备 ${eid} 不存在`);
    const row = M.db.insert("equipment_logs", {
      equipment_id: Number(eid),
      type: logType,
      content,
      operator: String(data.operator || "").trim(),
      cost,
      created_at: M.now(),
    });
    // 维修时把设备置为保养中，完成后用户可再手动改回
    if (logType === "repair") e.status = "maintenance";
    M.db.save();
    return logWithEquip(row);
  }

  // ── 路由（以 .py routes() 为准）─────────────────────────────────
  M.register("GET", "/api/equipment", () => ({ body: listEquipment() }));
  M.register("POST", "/api/equipment", (ctx) => ({ status: 201, body: createEquipment(ctx.body) }));
  M.register("POST", "/api/equipment/(?<id>\\d+)/status", (ctx) =>
    ({ body: setStatus(Number(ctx.params.id), (ctx.body && ctx.body.status) || "") }));
  M.register("GET", "/api/equipment/(?<id>\\d+)/logs", (ctx) =>
    ({ body: listLogs(Number(ctx.params.id)) }));
  M.register("POST", "/api/equipment/(?<id>\\d+)/logs", (ctx) =>
    ({ status: 201, body: addLog(Number(ctx.params.id), ctx.body) }));
  M.register("GET", "/api/equipment-logs", () => ({ body: listLogs() }));

  // ── 种子（与 .py seed 一致：3 台设备 + 首台一条点检记录）────────
  M.onSeed(function (M) {
    if (M.db.table("equipment").length) return;
    const e1 = createEquipment({
      code: "CNC-01", name: "数控车床", model: "CK6140",
      location: "一号车间", owner: "张工", status: "running",
    });
    createEquipment({
      code: "INJ-02", name: "注塑机", model: "HTF160",
      location: "二号车间", owner: "王工", status: "idle",
    });
    createEquipment({
      code: "WELD-03", name: "焊接机器人", model: "AR-1440",
      location: "三号车间", owner: "赵工", status: "fault",
    });
    addLog(e1.id, {
      type: "inspect", content: "日常点检：油位正常，导轨润滑良好",
      operator: "李师傅",
    });
  });
})();
